### ISSUE TITLE
Custom Access-Control-Allow Headers are not showing up

### ISSUE THREAD

**Issue by jachin** [aws-amplify/amplify-console#1563]
**Describe the bug**
I can not get my custom headers to show up.

I have tried several different things after my initial attempt failed:
 - In the build log I saw that having `customHeaders` in `amplify.yml` is deprecated so I moved them to `customHttp.yml` (the build found them and looks happy)
 - I thought maybe the order of the rules mattered, I tried putting the most "permissive" pattern at the bottom, that did not seem to help either.
 - I tried putting my `customHeaders.yml` in the web ui management console thing-y for Amplify too, no luck there either.

**To Reproduce**
1. Added customHeaders.yml
2. It looks something like:
    ```
    customHeaders:
    - pattern: '*.woff'
      headers:
        - key: 'Access-Control-Allow-Origin'
          value: 'https://www.mydomain.name'
        - key: 'Access-Control-Allow-Credentials'
          value: 'true'
        - key: 'Access-Control-Allow-Methods'
          value: 'GET, HEAD'
    - pattern: '*.woff2'
      headers:
        - key: 'Access-Control-Allow-Origin'
          value: 'https://www.mydomain.name'
        - key: 'Access-Control-Allow-Credentials'
          value: 'true'
        - key: 'Access-Control-Allow-Methods'
          value: 'GET, HEAD'
    - pattern: '**/*'
      headers:
        - key: 'X-Robots-Tag'
          value: 'noindex'
    ```

**Expected behavior**

For the woff and woff2 (font files) _have_ the custom headers, something like:

```
HTTP/2 200 OK
content-type: font/woff
content-length: 45516
last-modified: Fri, 12 Feb 2021 15:51:09 GMT
x-amz-server-side-encryption: AES256
accept-ranges: bytes
server: AmazonS3
x-robots-tag: noindex
Access-Control-Allow-Origin: https://www.mydomain.name
Access-Control-Allow-Methods: GET, HEAD
Access-Control-Allow-Credentials: true
date: Fri, 12 Feb 2021 20:06:13 GMT
etag: "82a9683860460a3aa1eadfc2e0180bc4"
x-cache: RefreshHit from cloudfront
via: 1.1 ac8301cb52f27894048b5f9019631931.cloudfront.net (CloudFront)
x-amz-cf-pop: MSP50-C2
x-amz-cf-id: YaST35fgSvBxnMXkARbJerHC1qPElCpDg75h349jGTYigWEiOG4vWw==
X-Firefox-Spdy: h2
```

But the responses are missing all the `Access-Control-Allow-*` headers

**Comment by VictorCcccc** [comment]
Hi @jachin , Can you try to change pattern to '**/*.woff' and '**/*.woff2' ?
Let me know have further question.

**Comment by jachin** [comment]
@VictorCcccc That worked!

In my case this worked out, but I feel like there's still a bug here, if `customHttp.yml` doesn't full support "globbing" (pattern matching) the docs should call that out.

**Comment by jachin** [comment]
@VictorCcccc Ok, so I said it worked... well... it worked on the branch "Preview". But I have sense merged that to master, and the headers are *not* showing up on the master build.

If I'm honest, it did not occur to me to test the preview builds for my other attempts so I'm sure if what I _had_ before was correctly setting the HTTP Headers in the preview builds already.

**Comment by jachin** [comment]
Oh man... this is starting to feel a little bit like gas lighting, but here's what I think I'm seeing.

This is on a "preview" build using the wildcards (in `customHttp.yml`) that look like `**/*.woff` and `**/*.woff2`.

When I request the `.woff` font, the response does *not* have the custom headers, but when I request the `.woff2` file that *does* have the custom headers.

However, when I hit the production URL, which has an identical `customHttp.yml` file, for both font file extensions, I am *not* getting the custom headers.

**Comment by VictorCcccc** [comment]
> Oh man... this is starting to feel a little bit like gas lighting, but here's what I think I'm seeing.
> 
> On a "preview" build using the wildcards (in `customHttp.yml`) that look like `**/*.woff` and `**/*.woff2`, when I request the `.woff` font, the response does _not_ have the custom headers, but when I request the `.woff2` file that _does_ have the custom headers.
> 
> However, when I request the same files in production (build off the `master` branch), which has an identical `customHttp.yml` file (as the preview branch), for both font file extensions, I am _not_ getting the custom headers.

Found this same issue when I reproduced your bug on my end. Still investigating this problem.

**Comment by VictorCcccc** [comment]
Hey @jachin, Can you send me your App ID to me? The App ID is a string start with "d" in your App ARN(after "/"). It will be helpful for us to locate the issue. Thanks!

**Comment by jachin** [comment]
> Hi @jachin, Can you send me your App ID? The App ID is a string start with "d" (after "/") in your App ARN. It will be helpful for us to locate the issue. Thanks!
> 
> The thing is that if there is a branch level config, that takes precedence over any app-level config. This may become the main reason why custom header show up "randomly". I'll check the config file for you on our back-end. And will request Docs updates!

It's `d11ckvg2f5ptem`, thanks for looking into it.

**Comment by VictorCcccc** [comment]
> > Hi @jachin, Can you send me your App ID? The App ID is a string start with "d" (after "/") in your App ARN. It will be helpful for us to locate the issue. Thanks!
> > The thing is that if there is a branch level config, that takes precedence over any app-level config. This may become the main reason why custom header show up "randomly". I'll check the config file for you on our back-end. And will request Docs updates!
> 
> It's `d11ckvg2f5ptem`, thanks for looking into it.

I've sent you (jachin.rupe@gmail.com) an email with details, Thanks.

**Comment by VictorCcccc** [comment]
Update: Found correct custom headers showed up on different browser. The missing custom header issue might be caused by the internal bug of browser engine.

**Comment by jachin** [comment]
> Update: Found correct custom headers showed up on different browser. The missing custom header issue might be caused by the internal bug of browser engine.

That's an interesting theory... I had not thought of that. I've tried a bunch of different "clients" though. Including Paw and curl. It's possible though these different clients are all using the same library or something...

Looking now though, at the production assets, I am seeing the correct headers! Hurray! I wonder if something did change on the Amplify side.

@VictorCcccc would you like me to try adding or editing a new custom header to see if the updated HTTP headers start working right away?

**Comment by VictorCcccc** [comment]
> > Update: Found correct custom headers showed up on different browser. The missing custom header issue might be caused by the internal bug of browser engine.
> 
> That's an interesting theory... I had not thought of that. When I was researching this bug I tried a bunch of different "clients" though. Including Paw and curl. It's possible though these different clients are all using the same library or something...
> 
> Looking now though, at the production assets, I am seeing the correct headers! Hurray! I wonder if something did change on the Amplify side.
> 
> @VictorCcccc would you like me to try adding new custom header or editing an existing to see if the updated HTTP headers start working right away?

I've sent you the details via email. Yes, you could add new header in customHttp.yml which is in your root folder. I tried to reproduce this error yesterday again and seems like that all the .woff2 file has correct header.

**Comment by jachin** [comment]
Just to close the loop on this. thanks again for all your help @VictorCcccc. The problem appears to be that the Cloud Front servers (The CDN that serves up the stuff on Amplify) doesn't actually _know_ when a new build has happened and if they should refresh their http headers (they do seem to know when the file has changed, but not the http headers). The work around was to _also_ set the `Cache-Control` header to _something_ in the `customHttp.yml` file and the CDN servers _should_ invalidate the cached data they use to make the http headers at least that often (there are [docs about this I skimmed](https://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/Expiration.html.)).

Hopefully this helps future travelers.

I suggested that the documentation for `customHttp.yml` in Amplify gets updated to explain this, and if possible. An even better solution would be to push to the CDNs something that told them to refresh their custom http headers when there was a successful build.

Anyway, I'm going to close this issue, I assume this is something the Amplify folks are tracking internally.
